### PR TITLE
Closes #576: Implement facades for importer DNet services

### LINKED ISSUE #576: Implement facades for dnet import services to simplify code and enable workflow integration tests
#553 follow-up.

### PR DESCRIPTION
**Pull request by marekhorst**

### CODE REVIEW

File: iis-wf/iis-wf-import/src/main/java/eu/dnetlib/iis/wf/importer/database/project/DatabaseServiceBasedProjectImporter.java [hunk 1]
```diff
@@ -75,70 +67,37 @@
...
+
+//			initializing database reader
+			DatabaseFacade databaseFacade = ServiceFacadeUtils.instantiate(parameters);
+
 //			reading sql query content
 			StringWriter writer = new StringWriter();
-			IOUtils.copy(this.getClass().getClassLoader()
-	                .getResourceAsStream("eu/dnetlib/iis/wf/importer/database/project/sql/read_project_details_v2.sql"), 
-	                writer, "utf-8");
+			IOUtils.copy(this.getClass().getClassLoader().getResourceAsStream(queryLocation), writer, "utf-8");
```
Comment (madryk) [inline iis-wf/iis-wf-import/src/main/java/eu/dnetlib/iis/wf/importer/database/project/DatabaseServiceBasedProjectImporter.java]: I would move reading this sql to private method returning already a string. 
And in this method wrap the writer in try block (it's a good convention to close opened writers - even if we know that they do nothing at closing)

File: iis-wf/iis-wf-import/src/main/java/eu/dnetlib/iis/wf/importer/dataset/DataciteMDStoreImporter.java [hunk 2]
```diff
@@ -79,66 +75,35 @@
...
-					new FileSystemPath(fs, portBindings.getOutput().get(PORT_OUT_DATASET_TO_MDSTORE)), 
-					DocumentToMDStore.SCHEMA$);
-			
+		FileSystem fs = FileSystem.get(conf);
+		
+        try (DataFileWriter<DataSetReference> datasetRefWriter = DataStore.create(
+                new FileSystemPath(fs, portBindings.getOutput().get(PORT_OUT_DATASET)), DataSetReference.SCHEMA$);
+                DataFileWriter<DocumentToMDStore> datasetToMDStoreWriter = DataStore.create(
+                        new FileSystemPath(fs, portBindings.getOutput().get(PORT_OUT_DATASET_TO_MDSTORE)),
+                        DocumentToMDStore.SCHEMA$)) {
```
Comment (madryk) [inline iis-wf/iis-wf-import/src/main/java/eu/dnetlib/iis/wf/importer/dataset/DataciteMDStoreImporter.java]: This initialize try resources block is messy. I would move initialization of `FileSystemPath`s to outside variables to keep this block simpler

File: iis-wf/iis-wf-import/src/main/java/eu/dnetlib/iis/wf/importer/dataset/DataciteMDStoreImporter.java [hunk 3]
```diff
@@ -79,66 +75,35 @@
...
-					rsFactory.setServiceResolver(new JaxwsServiceResolverImpl());
-					rsFactory.setPageSize(parameters.containsKey(IMPORT_DATACITE_MDSTORE_PAGESIZE)?
-									Integer.valueOf(parameters.get(IMPORT_DATACITE_MDSTORE_PAGESIZE)):
-										defaultPagesize);
-					SAXParserFactory parserFactory = SAXParserFactory.newInstance();
-					SAXParser saxParser = parserFactory.newSAXParser();
+				String[] mdStoreIds = StringUtils.split(mdStoresCSV, WorkflowRuntimeParameters.DEFAULT_CSV_DELIMITER);
+                
+				SAXParserFactory parserFactory = SAXParserFactory.newInstance();
+                SAXParser saxParser = parserFactory.newSAXParser();
```
Comment (madryk) [inline iis-wf/iis-wf-import/src/main/java/eu/dnetlib/iis/wf/importer/dataset/DataciteMDStoreImporter.java]: indentation problems

File: iis-wf/iis-wf-import/src/main/java/eu/dnetlib/iis/wf/importer/facade/DatabaseFacade.java [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+/**
+ * Database service facade.
+ * 
+ * @author mhorst
+ *
+ */
+public interface DatabaseFacade {
+
+    /**
+     * Delivers all records for given MDStore identifier
```
Comment (madryk) [inline iis-wf/iis-wf-import/src/main/java/eu/dnetlib/iis/wf/importer/facade/DatabaseFacade.java]: Invalid comment

File: iis-wf/iis-wf-import/src/main/java/eu/dnetlib/iis/wf/importer/facade/ObjectStoreFacade.java [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+/**
+ * ObjectStore service facade.
+ * 
+ * @author mhorst
+ *
+ */
+public interface ObjectStoreFacade {
+
+    /**
+     * Provides all metadata records from given objectstore.
```
Comment (madryk) [inline iis-wf/iis-wf-import/src/main/java/eu/dnetlib/iis/wf/importer/facade/ObjectStoreFacade.java]: From params we can deduce that not `all metadata` has to be provided.
I would add info about time range or remove word `all`

File: iis-wf/iis-wf-import/src/main/java/eu/dnetlib/iis/wf/importer/facade/ServiceFacadeUtils.java [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+    
+    /**
+     * Provides parameter value. Returns default value when entry not found among parameters.
+     * @param parameters map of parameters
+     * @param paramName parameter name
+     * @param defaultValue parameter default value to be returned when entry not found among parameters
+     */
+    public static String getValue(Map<String, String> parameters, String paramName, String defaultValue) {
+        return parameters.containsKey(paramName)?parameters.get(paramName):defaultValue;
+    }
```
Comment (madryk) [inline iis-wf/iis-wf-import/src/main/java/eu/dnetlib/iis/wf/importer/facade/ServiceFacadeUtils.java]: Probably it's not the best place to include this method. What do you think about extending WorkflowRuntimeParameters?

File: iis-wf/iis-wf-import/src/main/java/eu/dnetlib/iis/wf/importer/facade/WebServiceObjectStoreFacade.java [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+
+import org.apache.log4j.Logger;
+
+import eu.dnetlib.data.objectstore.rmi.ObjectStoreService;
+import eu.dnetlib.data.objectstore.rmi.ObjectStoreServiceException;
+import eu.dnetlib.enabling.resultset.client.ResultSetClientFactory;
+import eu.dnetlib.enabling.tools.JaxwsServiceResolverImpl;
+
+/**
+ * WebService based MDStore facade.
```
Comment (madryk) [inline iis-wf/iis-wf-import/src/main/java/eu/dnetlib/iis/wf/importer/facade/WebServiceObjectStoreFacade.java]: `ObjectStore facade`?

**Comment by madryk** [comment]
I would add unit test for `ServiceFacadeUtils` checking three cases: we don't provide parameter with factory name, provided parameter with factory name is not really a `ServiceFacadeFactory` and last one checking successful scenario.

File: iis-wf/iis-wf-import/src/main/java/eu/dnetlib/iis/wf/importer/facade/WebServiceDatabaseFacadeFactory.java [hunk 8]
```diff
@@ -0,0 +1,42 @@
+package eu.dnetlib.iis.wf.importer.facade;
+
+import static eu.dnetlib.iis.wf.importer.ImportWorkflowRuntimeParameters.IMPORT_DATABASE_SERVICE_LOCATION;
+import static eu.dnetlib.iis.wf.importer.ImportWorkflowRuntimeParameters.IMPORT_RESULT_SET_CLIENT_READ_TIMEOUT;
+import static eu.dnetlib.iis.wf.importer.ImportWorkflowRuntimeParameters.*;
```
Comment (madryk) [inline iis-wf/iis-wf-import/src/main/java/eu/dnetlib/iis/wf/importer/facade/WebServiceDatabaseFacadeFactory.java]: Organize imports?

File: iis-wf/iis-wf-import/src/main/java/eu/dnetlib/iis/wf/importer/facade/WebServiceDatabaseFacadeFactory.java [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+ *
+ */
+public class WebServiceDatabaseFacadeFactory implements ServiceFacadeFactory<DatabaseFacade> {
+
+    
+    private static final String IMPORT_DATABASE_CLIENT_CONNECTION_TIMEOUT = "import.database.client.connection.timeout";
+    
+    private static final String IMPORT_DATABASE_CLIENT_READ_TIMEOUT = "import.database.client.read.timeout";
+    
+    private static final String defaultDdatabaseConnectionTimeout = "60000";
```
Comment (madryk) [inline iis-wf/iis-wf-import/src/main/java/eu/dnetlib/iis/wf/importer/facade/WebServiceDatabaseFacadeFactory.java]: `Ddatabase` typo?

File: iis-wf/iis-wf-import/src/main/java/eu/dnetlib/iis/wf/importer/dataset/DataciteMDStoreImporter.java (hunk 3 above)
Comment (marekhorst) [inline iis-wf/iis-wf-import/src/main/java/eu/dnetlib/iis/wf/importer/dataset/DataciteMDStoreImporter.java]: Reformatted whole file.

File: iis-wf/iis-wf-import/src/main/java/eu/dnetlib/iis/wf/importer/facade/ObjectStoreFacade.java (hunk 5 above)
Comment (marekhorst) [inline iis-wf/iis-wf-import/src/main/java/eu/dnetlib/iis/wf/importer/facade/ObjectStoreFacade.java]: True, initialy I've skipped time range parameters and forgot to update comment after introducing `from` and `until` params. Comment updated.

File: iis-wf/iis-wf-import/src/main/java/eu/dnetlib/iis/wf/importer/database/project/DatabaseServiceBasedProjectImporter.java [hunk 10]
```diff
@@ -75,70 +68,45 @@
...
-		}	
+		}
+	}
+	
+    //------------------------ LOGIC --------------------------
+
+	/**
+	 * Reads query from classpath resource.
+	 */
+	private String loadQuery() throws IOException {
```
Comment (madryk) [inline iis-wf/iis-wf-import/src/main/java/eu/dnetlib/iis/wf/importer/database/project/DatabaseServiceBasedProjectImporter.java]: private section?

File: iis-wf/iis-wf-import/src/test/java/eu/dnetlib/iis/wf/importer/facade/ServiceFacadeUtilsTest.java [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+        String instantiated = ServiceFacadeUtils.instantiate(parameters);
+        
+        // assert
+        assertEquals(instantiated, StringFactory.class.getName());
+    }
+    
+    
+    // ------------------------ INNER CLASSES --------------------------
+    
+    public static class FactoryWithoutNoArgConstructor implements ServiceFacadeFactory<String> {
```
Comment (madryk) [inline iis-wf/iis-wf-import/src/test/java/eu/dnetlib/iis/wf/importer/facade/ServiceFacadeUtilsTest.java]: Factory most probably won't be used anywhere else except this test class so it should be private.

File: iis-wf/iis-wf-import/src/test/java/eu/dnetlib/iis/wf/importer/facade/ServiceFacadeUtilsTest.java [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+        }
+        
+        @Override
+        public String instantiate(Map<String, String> parameters) {
+            return this.getClass().getName();
+        }
+        
+    }
+    
+    public static class StringFactory implements ServiceFacadeFactory<String> {
```
Comment (madryk) [inline iis-wf/iis-wf-import/src/test/java/eu/dnetlib/iis/wf/importer/facade/ServiceFacadeUtilsTest.java]: same as above

**Comment by madryk** [comment]
Looks OK. I still have some small remarks. If you decide that they are legitimate then you can close this pull request

File: iis-wf/iis-wf-import/src/test/java/eu/dnetlib/iis/wf/importer/facade/ServiceFacadeUtilsTest.java (hunk 11 above)
Comment (marekhorst) [inline iis-wf/iis-wf-import/src/test/java/eu/dnetlib/iis/wf/importer/facade/ServiceFacadeUtilsTest.java]: True, had to explicitly add public constructor though to allow instantiation via reflection.

**Pull request closed without merging**